src/main/java/com/hubspot/ringleader/watcher/WatcherFactory.java

```diff
@@ -105,36 +111,55 @@ public void newEvent(Event event) {
     return curatorReference;
   }
 
-  void recordClose() {
+  void recordClose(PersistentWatcher watcher) {
     synchronized (watcherLock) {
-      if (watchers.decrementAndGet() == 0) {
+      watchers.remove(watcher);
+      if (watchers.size() == 0) {
         closed.set(true);
         executor.shutdown();
         synchronousCleanup(curatorReference.get());
       }
     }
   }
 
-  synchronized void replaceCurator(final Runnable runnable) {
+  synchronized void replaceCurator() {
     long timestamp = curatorTimestamp.get();
     long age = System.currentTimeMillis() - timestamp;
-    long minAge = TimeUnit.MINUTES.toMillis(1);
 
-    // only attempt reconnect once per minute so we don't exacerbate a failure scenario and don't reconnect when the
-    // executor is shutting down.
-    if (age < minAge || !curatorTimestamp.compareAndSet(timestamp, System.currentTimeMillis()) || executor.isShutdown()) {
+    if (executor.isShutdown()) {
+      // don't reconnect when the executor is shutting down.
       return;
     }
 
+    if (replacementScheduled.get()) {
+      return;
+    }
+
+    // only attempt reconnect once per replacementIntervalMillis so we don't exacerbate a failure scenario
+    if (age < replacementIntervalMillis) {
+      if (!replacementScheduled.get()) {
+        long millisToWait = replacementIntervalMillis - age;
```

Comment:
Rather than dropping repeated calls within a minute, this just delays the replacement. This ensures we'll always reconnect.